src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.List.cs

```diff
@@ -0,0 +1,374 @@
+        {
+            Debug.Assert(_list != null);
+            Debug.Assert(_stringComparer != null);
+
+            for (int i = 0; i < _list.Count; i++)
+            {
+                KeyValuePair<string, JsonNode?> current = _list[i];
+                if (_stringComparer.Equals(propertyName, current.Key))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private void CreateDictionaryIfThreshold()
+        {
+            Debug.Assert(_list != null);
+            Debug.Assert(_stringComparer != null);
+
+            if (_dictionary == null && _list.Count > ListToDictionaryThreshold)
+            {
+                _dictionary = JsonHelpers.CreateDictionaryFromCollection(_list, _stringComparer);
+            }
+        }
+
+        private void ClearNodes()
+        {
+            if (_jsonElement != null)
+            {
+                Debug.Assert(_list == null);
+                Debug.Assert(_dictionary == null);
+                _jsonElement = null;
+                return;
+            }
+
+            foreach (JsonNode? node in GetValueCollection(this))
+            {
+                DetachParent(node);
+            }
+
+            _list?.Clear();
+            _dictionary?.Clear();
+            ClearLastValueCache();
+        }
+
+        /// <summary>
+        /// This is based on ICollection.Contains(KeyValuePair) and is explicitely implemented and
+        /// is not expected to be used often since IDictionary.ContainsKey(key) is more useful.
+        /// </summary>
+        private bool ContainsNode(KeyValuePair<string, JsonNode?> node)
+        {
+            foreach (KeyValuePair<string, JsonNode?> item in this)
+            {
+                Debug.Assert(_stringComparer != null);
+                if (ReferenceEquals(item.Value, node.Value) && _stringComparer.Equals(item.Key, node.Key))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool ContainsNode(JsonNode? node)
+        {
+            CreateList();
+
+            foreach (JsonNode? item in GetValueCollection(this))
+            {
+                if (ReferenceEquals(item, node))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private KeyValuePair<string, JsonNode?>? FindNode(JsonNode? node)
+        {
+            CreateList();
+
+            foreach (KeyValuePair<string, JsonNode?> item in this)
+            {
+                if (ReferenceEquals(item.Value, node))
+                {
+                    return item;
+                }
+            }
+
+            return null;
+        }
+
+        private bool ContainsNode(string propertyName)
+        {
+            if (_dictionary != null)
+            {
+                return _dictionary.ContainsKey(propertyName);
+            }
```

Comment:
Doesn't seem like the dictionary can be `null` if this method is called.

Reply:
Also, if this cannot be changed into an assert perhaps rewrite to `return _dictionary?.ContainsKey(propertyName);`

Reply:
_dictionary can be null here whenever the threshold isn't hit yet. E.g.:
```cs
JsonObject jObject = JsonNode.Parse("{}").AsObject();
bool exists = jObject.ContainsKey("FOO");
```